### PR TITLE
[RND-21] Make descriptors more of a 1st class resource

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: Meadowlark-js/packages/meadowlark-core/src/handler/DescriptorLoader.ts [hunk 1]
```diff
@@ -116,23 +111,15 @@ async function loadParsedDescriptors(descriptorData: XmlDescriptorData): Promise
...
-      const documentInfo: DocumentInfo = {
-        ...newDocumentInfo(),
-        documentIdentity: [
-          {
-            name: 'descriptor',
-            value: `${descriptorDocument.namespace}#${descriptorDocument.codeValue}`,
-          },
-        ],
-      };
+      const descriptorDocument: DescriptorDocument = decapitalizeKeys(descriptorReference) as DescriptorDocument;
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/DescriptorLoader.ts:121]: DescriptorLoader is the only place the descriptor document shape and identity was defined. Pulled out a few pieces to share.

File: Meadowlark-js/packages/meadowlark-core/src/model/DescriptorDocument.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+export interface DescriptorDocument {
+  namespace: string;
+  codeValue: string;
+  shortDescription: string;
+  description?: string;
+}
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/model/DescriptorDocument.ts:11]: I noticed there are other possible fields on an ODS/API descriptor document, but didn't look at expanding.

File: Meadowlark-js/packages/meadowlark-core/src/model/DescriptorDocumentInfo.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/model/DescriptorDocumentInfo.ts:5]: Lifted from descriptor loading

File: Meadowlark-js/packages/meadowlark-core/src/validation/DocumentIdentityExtractor.ts [hunk 4]
```diff
@@ -95,11 +97,20 @@ function documentIdentitiesFrom(
...
+  return { documentIdentity: descriptorDocumentIdentityFrom(body), studentId: null, edOrgId: null };
+}
+
 /**
  * Takes a MetaEd entity object and a API JSON body for the resource mapped to that MetaEd entity and
  * extracts the document identity information from the JSON body. Also extracts security information, if any.
  */
 export function extractDocumentIdentity(entity: TopLevelEntity, body: object): DocumentIdentityWithSecurity {
+  if (entity.type === 'descriptor') return descriptorDocumentIdentityWithSecurity(body as DescriptorDocument);
+
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/validation/DocumentIdentityExtractor.ts:113]: If it's a descriptor, the body's been schema validated at this point, we know it's a DescriptorDocument

File: Meadowlark-js/packages/meadowlark-core/src/validation/DocumentIdentityExtractor.ts [hunk 5]
```diff
@@ -95,11 +97,20 @@ function documentIdentitiesFrom(
   return result;
 }
 
+/**
+ * All descriptor documents have the same identity fields
+ */
+function descriptorDocumentIdentityWithSecurity(body: DescriptorDocument): DocumentIdentityWithSecurity {
+  return { documentIdentity: descriptorDocumentIdentityFrom(body), studentId: null, edOrgId: null };
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/validation/DocumentIdentityExtractor.ts:104]: Use the identity function we split out

File: Meadowlark-js/packages/meadowlark-core/src/validation/DocumentValidator.ts [hunk 6]
```diff
@@ -67,7 +67,7 @@ export async function validateDocument(
   const documentReferences: DocumentReference[] = [];
   const descriptorReferences: DocumentReference[] = [];
 
-  if (!resourceInfo.isDescriptor && body != null) {
+  if (body != null) {
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/validation/DocumentValidator.ts:70]: Now we can handle descriptors here

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiEntityMappingEnhancer.ts [hunk 7]
```diff
@@ -42,6 +42,11 @@ export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
     (entity.data.meadowlark as EntityMeadowlarkData).apiMapping = buildApiEntityMapping(entity as TopLevelEntity);
   });
 
+  // Descriptors have no API shape metadata
+  getAllEntitiesOfType(metaEd, 'descriptor').forEach((entity) => {
+    (entity.data.meadowlark as EntityMeadowlarkData).apiMapping = NoApiEntityMapping;
+  });
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/ApiEntityMappingEnhancer.ts:48]: Descriptors just need the null object version of ApiEntityMapping

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 8]
```diff
@@ -21,6 +21,14 @@ import { singularize, topLevelNameOnEntity } from '../Utility';
...
 const enhancerName = 'JoiSchemaEnhancer';
 
+// All descriptor documents have the same schema
+const descriptorSchema = Joi.object({
+  namespace: Joi.string().required(),
+  codeValue: Joi.string().required(),
+  shortDescription: Joi.string().required(),
+  description: Joi.string().optional(),
+});
+
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:31]: If we want to expand the valid fields on descriptor documents, this is the place to do it.

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts [hunk 9]
```diff
@@ -228,13 +236,20 @@ function buildJoiSchema(entityForSchema: TopLevelEntity): Joi.AnySchema {
...
       entityMeadowlarkData.joiSchema = buildJoiSchema(entity as TopLevelEntity);
     },
   );
 
+  // Attach descriptor schema to each descriptor
+  getAllEntitiesOfType(metaEd, 'descriptor').forEach((entity) => {
+    const entityMeadowlarkData = entity.data.meadowlark as EntityMeadowlarkData;
+    entityMeadowlarkData.joiSchema = descriptorSchema;
+  });
+
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:252]: Descriptors all have the same schema

Comment (bradbanister) [reply Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/enhancer/JoiSchemaEnhancer.ts:252]: As an aside, everywhere you see `'descriptor'` in this PR, it's a Typescript string literal type that enumerates the valid MetaEd types.

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/model/ApiEntityMapping.ts [hunk 10]
```diff
@@ -54,5 +54,4 @@ export function newApiEntityMapping(): ApiEntityMapping {
 
 export const NoApiEntityMapping: ApiEntityMapping = deepFreeze({
   ...newApiEntityMapping(),
-  metaEdName: 'NoApiEntityMapping',
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/model/ApiEntityMapping.ts:57]: No idea why this is here. `ApiEntityMapping` doesn't have a `metaEdName`.

File: Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/model/EntityMeadowlarkData.ts [hunk 11]
```diff
@@ -40,12 +40,17 @@ export function addEntityMeadowlarkDataTo(entity: ModelBase) {
...
-    },
-  );
+  getAllEntitiesOfType(
+    metaEd,
+    'domainEntity',
+    'association',
+    'domainEntitySubclass',
+    'associationSubclass',
+    'descriptor',
+  ).forEach((entity) => {
```
Comment (bradbanister) [inline Meadowlark-js/packages/metaed-plugin-edfi-meadowlark/src/model/EntityMeadowlarkData.ts:50]: This is an initializer for the Meadowlark entity metadata object. Now that descriptors have metadata, they need to be initialized with it. Prettier made this look bigger than it is.

**Review by stephenfuqua — APPROVED** [review]

**Pull request merged**